File: compiler/feedback/structures.go
package feedback

import (
	"wipple/database"
	"wipple/queries"
)

func registerStructures() {
	register(Feedback[string]{
		Id:    "missing-field",
		Rank:  RankSyntax,
		Query: queries.MissingField,
		Render: func(render *Render, node database.Node, field string) {
			render.WriteNode(node)
			render.WriteString(" is missing a pattern for the field ")
			render.WriteCode(field)
			render.WriteString(".")
			render.WriteBreak()
			render.WriteString("Try adding a pattern for this field using ")
			render.WriteCode(":")
			render.WriteString(".")
		},
	})

	register(Feedback[string]{
		Id:    "extra-field",
		Rank:  RankSyntax,
		Query: queries.ExtraField,
		Render: func(render *Render, node database.Node, field string) {
			render.WriteCode(field)
			render.WriteString(" isn't a field on this structure.")
			render.WriteBreak()
			render.WriteString("Double-check your spelling.")
		},
	})

	register(Feedback[struct{}]{
		Id:    "duplicate-field",
		Rank:  RankSyntax,
		Query: queries.DuplicateField,
		Render: func(render *Render, node database.Node, data struct{}) {
			render.WriteNode(node)
			render.WriteString(" is defined more than once.")
			render.WriteBreak()
			render.WriteString("Try removing this field.")
		},
	})
}

File: compiler/src/feedback/items/structures.rs
use crate::{
    feedback::{FeedbackCtx, FeedbackRank, RegisteredFeedback},
    nodes::{DuplicateField, ExtraField, MissingField},
    queries,
};
use std::collections::BTreeSet;

pub fn register(ctx: &mut FeedbackCtx) {
    ctx.register(RegisteredFeedback::new(
        "missing-field",
        queries::fact::<MissingField>,
        |_| FeedbackRank::Syntax,
        |(node, _)| (node.clone(), BTreeSet::new()),
        |w, (node, MissingField(field))| {
            w.write_node(node);
            w.write_string(" is missing a pattern for the field ");
            w.write_code(field);
            w.write_string(".");
            w.write_break();
            w.write_string("Try adding a pattern for this field using ");
            w.write_code(":");
            w.write_string(".");
        },
    ));

    ctx.register(RegisteredFeedback::new(
        "extra-field",
        queries::fact::<ExtraField>,
        |_| FeedbackRank::Syntax,
        |(node, _)| (node.clone(), BTreeSet::new()),
        |w, (_, ExtraField(field))| {
            w.write_code(field);
            w.write_string(" isn't a field on this structure.");
            w.write_break();
            w.write_string("Double-check your spelling.");
        },
    ));

    ctx.register(RegisteredFeedback::new(
        "duplicate-field",
        queries::fact::<DuplicateField>,
        |_| FeedbackRank::Syntax,
        |(node, _)| (node.clone(), BTreeSet::new()),
        |w, (node, _)| {
            w.write_node(node);
            w.write_string(" is defined more than once.");
            w.write_break();
            w.write_string("Try removing this field.");
        },
    ));
}
